flink-connectors/flink-connector-hbase-2.2/src/test/java/org/apache/flink/connector/hbase2/source/HBaseRowDataAsyncLookupFunctionTest.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.hbase2.source;
+
+import org.apache.flink.connector.hbase.options.HBaseLookupOptions;
+import org.apache.flink.connector.hbase.util.HBaseTableSchema;
+import org.apache.flink.connector.hbase.util.PlannerType;
+import org.apache.flink.connector.hbase2.util.HBaseTestBase;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.table.data.RowData;
+
+import org.apache.flink.shaded.guava18.com.google.common.collect.Lists;
+
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static org.junit.Assert.assertEquals;
+
+/** Test suite for {@link HBaseRowDataAsyncLookupFunction}. */
+public class HBaseRowDataAsyncLookupFunctionTest extends HBaseTestBase {
+
+    @Override
+    protected PlannerType planner() {
+        // lookup table source is only supported in blink planner
+        return PlannerType.BLINK_PLANNER;
+    }
+
+    @Test
+    public void testEval() throws Exception {
+        HBaseRowDataAsyncLookupFunction lookupFunction = buildRowDataAsyncLookupFunction();
+
+        lookupFunction.open(null);
+        List<RowData> list = new ArrayList<>();
+        int[] rowKey = {1, 2, 3};
+        for (int i = 0; i < rowKey.length; i++){
+            CompletableFuture<Collection<RowData>> feature = new CompletableFuture<>();
+            lookupFunction.eval(feature, rowKey[i]);
+            list.add(feature.get().iterator().next());
+        }
+        lookupFunction.close();
+        List<String> result =
+            Lists.newArrayList(list).stream()
+                .map(RowData::toString)
+                .sorted()
+                .collect(Collectors.toList());
```

Comment:
This makes the async lookup to be sync and doesn't test the feature. I suggest to change it to async fashion, and it reveals a bug in HBaseSerDe. You can use the following test:


```java
final List<String> result = new ArrayList<>();
        int[] rowkeys = {1, 2, 1, 12, 3, 12, 4, 3};
        CountDownLatch latch = new CountDownLatch(rowkeys.length);
        for (int rowkey : rowkeys) {
            CompletableFuture<Collection<RowData>> future = new CompletableFuture<>();
            lookupFunction.eval(future, rowkey);
            future.whenComplete(
                    (rs, t) -> {
                        System.out.println("complete: " + rs);
                        synchronized (result) {
                            if (rs.isEmpty()) {
                                result.add(rowkey + ": null");
                            } else {
                                rs.forEach(row -> result.add(rowkey + ": " + row.toString()));
                            }
                        }
                        latch.countDown();
                    });
        }

        // this verifies lookup calls are async
        assertTrue(result.size() < rowkeys.length);
        latch.await();
        
        assertEquals(...);
```

Reply:
The above test also tests non-exist keys, and please also test it using a no-cache async lookup function.